### PR TITLE
feat: Added HPO example

### PR DESCRIPTION
**Pull request by SiddhantSadangi**
# Description

Added HPO example

__Related to:__ https://app.clickup.com/t/86965qmcd

__Any expected test failures?__

---

Add a `[X]` to relevant checklist items

## ❔ This change

- [X] adds a new feature
- [ ] fixes breaking code
- [ ] is cosmetic (refactoring/reformatting)

---

## ✔️ Pre-merge checklist

- [X] Refactored code ([sourcery](https://sourcery.ai/))
- [X] Tested code locally
- [X] Precommit installed and run before pushing changes
- [ ] Added code to GitHub tests ([notebooks](workflows/test-notebooks.yml), [scripts](workflows/test-scripts.yml))
- [X] Updated GitHub [README](../README.md)
- [ ] Updated the projects overview page on Notion

---

## 🧪 Test Configuration

- OS: Windows11
- Python version: 3.10.11
- Neptune version: neptune-scale==0.6.2
- Affected libraries with version:

### CODE REVIEW

File: how-to-guides/hpo/notebooks/Neptune_HPO.ipynb [hunk 1]
```diff
@@ -0,0 +1,647 @@
...
+   "source": [
+    "### Explore the results in Neptune\n",
+    "Follow the link to the run and explore the logged metadata (such as metrics and hyperparameters) in Neptune:\n",
+    "\n",
+    "- The best trial, with its metrics and parameters, is available in the *best* namespace\n",
+    "- Metadata across all trials is available in the *trials* namespace\n",
+    "\n",
+    "To organize all relevant metadata in one view, create a [custom dashboard](https://docs-beta.neptune.ai/custom_dashboard). [See an example](https://scale.neptune.ai/o/examples/org/hpo/runs/details?viewId=9d4424ec-5c27-4933-9003-d62e0784ac68&detailsTab=dashboard&dashboardId=HPO-overview-9d4421e6-dfe4-400b-9dfb-d9b9e8a416b6&runIdentificationKey=HPO-11&type=run).\n",
+    "\n",
+    "You can also create [saved table views](https://docs-beta.neptune.ai/experiments_table#custom-views) to view best trials across different runs. An example is available [here](https://scale.neptune.ai/o/examples/org/hpo/runs/table?viewId=9d4424ec-5c27-4933-9003-d62e0784ac68&detailsTab=dashboard&dash=table&type=run)."
```
Comment (szaganek) [inline how-to-guides/hpo/notebooks/Neptune_HPO.ipynb:403]:
```suggestion
    "You can also create [saved table views](https://docs-beta.neptune.ai/experiments_table#custom-views) to view best trials across different runs. [See an example](https://scale.neptune.ai/o/examples/org/hpo/runs/table?viewId=9d4424ec-5c27-4933-9003-d62e0784ac68&detailsTab=dashboard&dash=table&type=run)."
```

File: how-to-guides/hpo/notebooks/Neptune_HPO.ipynb [hunk 2]
```diff
@@ -0,0 +1,647 @@
...
+    "sweep_run.close()"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Explore the results in Neptune\n",
+    "Follow the link to the runs and explore the logged metadata (such as metrics and hyperparameters) in Neptune:\n",
```
Comment (szaganek) [inline how-to-guides/hpo/notebooks/Neptune_HPO.ipynb:612]:
```suggestion
    "Follow the link to the runs and explore the logged metadata such as metrics and hyperparameters in the Neptune app:\n",
```

File: how-to-guides/hpo/notebooks/Neptune_HPO.ipynb [hunk 3]
```diff
@@ -0,0 +1,647 @@
...
+    "\n",
+    "- **Multiple runs**\n",
+    "  - To group all trials under a sweep, use the [run groups](https://docs-beta.neptune.ai/groups). [See an example](https://scale.neptune.ai/o/examples/org/hpo/runs/table?viewId=9d44261f-32a1-42e7-96ff-9b35edc4be66&detailsTab=dashboard&dash=table&type=run).\n",
+    "  - To compare trails within or across sweeps, create a [multi-run dashboard](https://docs-beta.neptune.ai/custom_dashboard#multi-run-dashboard). [See an example](https://scale.neptune.ai/o/examples/org/hpo/runs/compare?viewId=9d44261f-32a1-42e7-96ff-9b35edc4be66&detailsTab=dashboard&dash=dashboard&dashboardId=Compare-trials-9d44284a-40fe-4614-a66d-a5ca81b8b4cd&type=run&compare=uIWrlI2f5Tyn_lrTzrCY6RSrOVUYtMkY0ozkGXHFv6E8). \n",
+    "    - Turn on [*Average grouped runs*](https://docs-beta.neptune.ai/charts#comparing-grouped-runs) in the chart widget settings to compare the average of trials across different sweeps.\n",
+    "  - Export charts/dashboards to a [report](https://docs-beta.neptune.ai/reports) to see both sweep level and trial level comparisons together. [See an example](https://scale.neptune.ai/o/examples/org/hpo/reports/9d442900-19b4-47dc-a2e9-0faedc1f4d2c)"
```
Comment (szaganek) [inline how-to-guides/hpo/notebooks/Neptune_HPO.ipynb:622]:
```suggestion
    "  - Export charts/dashboards to a [report](https://docs-beta.neptune.ai/reports) to see both sweep level and trial level comparisons together. [See an example](https://scale.neptune.ai/o/examples/org/hpo/reports/9d442900-19b4-47dc-a2e9-0faedc1f4d2c)."
```

**Review by LeoRoccoBreedt — DISMISSED** [review]

File: how-to-guides/hpo/notebooks/Neptune_HPO.ipynb [hunk 4]
```diff
@@ -0,0 +1,647 @@
...
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Introduction\n",
+    "\n",
+    "When running a hyperparameter optimization job, you can use Neptune Scale to track all the metadata from the study and each trial.\n",
+    "\n",
+    "In this guide, you'll learn how to configure Neptune Scale to track the metadata of your hyperparameter optimization job."
```
Comment (normandy7) [inline how-to-guides/hpo/notebooks/Neptune_HPO.ipynb:32]: The previous sentence conveys this already.

```suggestion
```

File: how-to-guides/hpo/notebooks/Neptune_HPO.ipynb [hunk 5]
```diff
@@ -0,0 +1,647 @@
...
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from random import random\n",
+    "\n",
+    "run = Run(\n",
+    "    family=\"hpo\",\n",
+    "    run_id=f\"hpo-{random()}\",\n",
+    ")\n",
```
Comment (normandy7) [inline how-to-guides/hpo/notebooks/Neptune_HPO.ipynb:281]: Experiment name?

File: how-to-guides/hpo/notebooks/Neptune_HPO.ipynb [hunk 6]
```diff
@@ -0,0 +1,647 @@
...
+   "source": [
+    "run.close()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Explore the results in Neptune\n",
+    "Follow the link to the run and explore the logged metadata such as metrics and hyperparameters in the Neptune app:\n",
```
Comment (normandy7) [inline how-to-guides/hpo/notebooks/Neptune_HPO.ipynb:396]:
```suggestion
    "Follow the link to the run and explore the logged metadata in the Neptune app:\n",
```

File: how-to-guides/hpo/notebooks/Neptune_HPO.ipynb [hunk 7]
```diff
@@ -0,0 +1,647 @@
...
+    "\n",
+    "- The best trial, with its metrics and parameters, is available in the *best* namespace\n",
+    "- Metadata across all trials is available in the *trials* namespace\n",
+    "\n",
+    "To organize all relevant metadata in one view, create a [custom dashboard](https://docs-beta.neptune.ai/custom_dashboard). [See an example](https://scale.neptune.ai/o/examples/org/hpo/runs/details?viewId=9d4424ec-5c27-4933-9003-d62e0784ac68&detailsTab=dashboard&dashboardId=HPO-overview-9d4421e6-dfe4-400b-9dfb-d9b9e8a416b6&runIdentificationKey=HPO-11&type=run).\n",
+    "\n",
+    "You can also create [saved table views](https://docs-beta.neptune.ai/experiments_table#custom-views) to view best trials across different runs. [See an example](https://scale.neptune.ai/o/examples/org/hpo/runs/table?viewId=9d4424ec-5c27-4933-9003-d62e0784ac68&detailsTab=dashboard&dash=table&type=run)."
```
Comment (normandy7) [inline how-to-guides/hpo/notebooks/Neptune_HPO.ipynb:403]:
```suggestion
    "To view best trials across different runs, you can also create [saved table views](https://docs-beta.neptune.ai/experiments_table#custom-views). [See an example](https://scale.neptune.ai/o/examples/org/hpo/runs/table?viewId=9d4424ec-5c27-4933-9003-d62e0784ac68&detailsTab=dashboard&dash=table&type=run)."
```

File: how-to-guides/hpo/notebooks/Neptune_HPO.ipynb [hunk 8]
```diff
@@ -0,0 +1,647 @@
...
+    "sweep_run.close()"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Explore the results in Neptune\n",
+    "Follow the link to the runs and explore the logged metadata such as metrics and hyperparameters in the Neptune app:\n",
```
Comment (normandy7) [inline how-to-guides/hpo/notebooks/Neptune_HPO.ipynb:612]:
```suggestion
    "Follow the link to the runs and explore the logged metadata in the Neptune app:\n",
```

File: how-to-guides/hpo/notebooks/Neptune_HPO.ipynb [hunk 9]
```diff
@@ -0,0 +1,647 @@
...
+    "- **Multiple runs**\n",
+    "  - To group all trials under a sweep, use the [run groups](https://docs-beta.neptune.ai/groups). [See an example](https://scale.neptune.ai/o/examples/org/hpo/runs/table?viewId=9d44261f-32a1-42e7-96ff-9b35edc4be66&detailsTab=dashboard&dash=table&type=run).\n",
+    "  - To compare trails within or across sweeps, create a [multi-run dashboard](https://docs-beta.neptune.ai/custom_dashboard#multi-run-dashboard). [See an example](https://scale.neptune.ai/o/examples/org/hpo/runs/compare?viewId=9d44261f-32a1-42e7-96ff-9b35edc4be66&detailsTab=dashboard&dash=dashboard&dashboardId=Compare-trials-9d44284a-40fe-4614-a66d-a5ca81b8b4cd&type=run&compare=uIWrlI2f5Tyn_lrTzrCY6RSrOVUYtMkY0ozkGXHFv6E8). \n",
+    "    - Turn on [*Average grouped runs*](https://docs-beta.neptune.ai/charts#comparing-grouped-runs) in the chart widget settings to compare the average of trials across different sweeps.\n",
+    "  - Export charts/dashboards to a [report](https://docs-beta.neptune.ai/reports) to see both sweep level and trial level comparisons together. [See an example](https://scale.neptune.ai/o/examples/org/hpo/reports/9d442900-19b4-47dc-a2e9-0faedc1f4d2c)."
```
Comment (normandy7) [inline how-to-guides/hpo/notebooks/Neptune_HPO.ipynb:622]:
```suggestion
    "    - To compare the average of trials across different sweeps, turn on [*Average grouped runs*](https://docs-beta.neptune.ai/charts#comparing-grouped-runs) in the chart widget settings.\n",
    "  - To see both sweep-level and trial-level comparisons together, export charts or dashboards to a [report](https://docs-beta.neptune.ai/reports). [See an example](https://scale.neptune.ai/o/examples/org/hpo/reports/9d442900-19b4-47dc-a2e9-0faedc1f4d2c)."
```

File: how-to-guides/hpo/notebooks/Neptune_HPO.ipynb (hunk 5 above)
Comment (SiddhantSadangi) [reply how-to-guides/hpo/notebooks/Neptune_HPO.ipynb:281]: Experiments not needed for this use-case

File: how-to-guides/hpo/notebooks/Neptune_HPO.ipynb [hunk 10]
```diff
@@ -28,8 +28,6 @@
     "## Introduction\n",
     "\n",
     "When running a hyperparameter optimization job, you can use Neptune Scale to track all the metadata from the study and each trial.\n",
```
Comment (SiddhantSadangi) [inline how-to-guides/hpo/notebooks/Neptune_HPO.ipynb:30]:
```suggestion
    "When running a hyperparameter optimization job, you can use Neptune Scale to track all the metadata from the study and each trial.\n"
```

File: how-to-guides/hpo/notebooks/Neptune_HPO.ipynb (hunk 5 above)
Comment (normandy7) [reply how-to-guides/hpo/notebooks/Neptune_HPO.ipynb:279]: Hmm, alright. I was just under the impression that we should treat experiments as the main entity and runs as little more than "experiment remnants". But let's maybe revisit this once we have the new data model.

**Review by normandy7 — APPROVED** [review]

**Pull request merged**